src/bson_mapper/bson_archiver.hpp

```diff
@@ -512,6 +521,54 @@ class BSONInputArchive : public cereal::InputArchive<BSONInputArchive> {
 
    public:
     /**
+     * Checks if the next invocation of search() will yield a value. Used to check if a particular
+     * optional element, embedded document, or embedded array exists.
+     *
+     * If search() would indeed return a value, it is cached here so that the logic in search()
+     * will not need to be repeated.
+     *
+     * @return true if the next invocation of search with the current _nextName yields a value,
+     *         false otherwise.
+     */
+    bool willSearchYieldValue() {
+        // Make sure the node stack is not empty.
+        if (_nodeTypeStack.empty()) {
+            throw bson_mapper::Exception("Cannot search for element if not in node.");
+        }
+
+        // Check to make sure a document has been read.
+        if (!_readFirstDoc) {
+            throw bson_mapper::Exception(
+                "Error in element search, never loaded BSON data from file.");
+        }
+
+        if (_nextName) {
+            const char* nextName = _nextName;
+            _nextName = nullptr;
+
+            if (_nodeTypeStack.top() == InputNodeType::InObject ||
+                _nodeTypeStack.top() == InputNodeType::InRootElement) {
+                const auto& val = _curBsonDoc[nextName];
+                if (val) {
+                    _cachedSearchResult = val.get_value();
+                    return true;
+                }
+            } else if (_nodeTypeStack.top() == InputNodeType::InEmbeddedObject) {
+                const auto& val = _embeddedBsonDocStack.top()[nextName];
+                if (val) {
+                    _cachedSearchResult = val.get_value();
+                    return true;
+                }
+            }
+        } else if (_nodeTypeStack.top() == InputNodeType::InEmbeddedArray) {
```

Comment:
For these three cases, I think it might be cleaner to:
- check for array first and then throw (for less nesting)
- declare `val`, then assign conditionally on the type stack top
- un-duplicate the caching / return lines